presto-main/src/main/java/com/facebook/presto/operator/exchange/LocalMergeExchange.java

```diff
@@ -0,0 +1,178 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.exchange;
+
+import com.facebook.presto.spi.Page;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Queue;
+import java.util.stream.IntStream;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkElementIndex;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
+public class LocalMergeExchange
+{
+    private static final int DEFAULT_MAX_BUFFERED_PAGES = 2;
+
+    private final List<ExchangeBuffer> buffers;
+
+    public LocalMergeExchange(int buffersCount)
+    {
+        this(buffersCount, DEFAULT_MAX_BUFFERED_PAGES);
+    }
+
+    public LocalMergeExchange(int buffersCount, int maxBufferedPagesPerBuffer)
+    {
+        checkArgument(buffersCount > 0, "buffersCount must be greater than zero");
+        checkArgument(maxBufferedPagesPerBuffer > 0, "maxBufferedPagesPerBuffer must be greater than zero");
+        buffers = IntStream.range(0, buffersCount)
+                .mapToObj((i) -> new ExchangeBuffer(maxBufferedPagesPerBuffer))
+                .collect(toImmutableList());
+    }
+
+    public List<ExchangeBuffer> getBuffers()
+    {
+        return buffers;
+    }
+
+    public ExchangeBuffer getBuffer(int index)
+    {
+        checkElementIndex(index, buffers.size());
+        return buffers.get(index);
+    }
+
+    public void finishRead()
+    {
+        buffers.forEach(ExchangeBuffer::finishRead);
+    }
+
+    public void close()
+    {
+        finishRead();
+    }
+
+    public static class ExchangeBuffer
```

Comment:
>  Is it because you want to ensure that there is at least one input from each source

Yes, it was to hard to reuse those bits of code. Those are intended to be used when the number of sources is unknown.

> I wonder if the current approach, with a default max buffered pages of 2, would cause a lot of blocking, though

The exchange is inmemory. I was even thinking of buffering only single page. Pages are fairly big by comparing to the cost of blocking to exchange.